diesel/src/sqlite/types/json.rs

```diff
@@ -0,0 +1,852 @@
+    let float_value = float_str
+        .parse::<f64>()
+        .map_err(|_| "Failed to parse JSONB float")?;
+    Ok(serde_json::Value::Number(
+        serde_json::Number::from_f64(float_value).ok_or("Invalid float value")?,
+    ))
+}
+
+// Read a JSON string
+fn read_jsonb_text(bytes: &[u8], payload_size: usize) -> deserialize::Result<serde_json::Value> {
+    let text_bytes = &bytes[..payload_size];
+    let text = std::str::from_utf8(text_bytes).map_err(|_| "Invalid UTF-8 in JSONB string")?;
+    Ok(serde_json::Value::String(text.to_string()))
+}
+
+fn read_jsonb_textj(bytes: &[u8], payload_size: usize) -> deserialize::Result<serde_json::Value> {
+    let text_bytes = &bytes[..payload_size];
+    let text = std::str::from_utf8(text_bytes).map_err(|_| "Invalid UTF-8 in JSONB string")?;
+
+    // Unescape JSON escape sequences (e.g., "\n", "\u0020")
+    let unescaped_text = serde_json::from_str(&format!("\"{}\"", text))
+        .map_err(|_| "Failed to parse JSON-escaped text in TEXTJ")?;
+
+    Ok(unescaped_text)
+}
+
+// Read a JSON array
+fn read_jsonb_array(bytes: &[u8], payload_size: usize) -> deserialize::Result<serde_json::Value> {
+    let mut elements = Vec::new();
+    let mut total_read = 0;
+
+    while total_read < payload_size {
+        let (element, consumed) = read_jsonb_value(&bytes[total_read..payload_size])?;
+
+        elements.push(element);
+        total_read += consumed;
+    }
+
+    if total_read != payload_size {
+        return Err("Array payload size mismatch".into());
+    }
+
+    Ok(serde_json::Value::Array(elements))
+}
+
+fn read_jsonb_object(bytes: &[u8], payload_size: usize) -> deserialize::Result<serde_json::Value> {
+    let mut object = serde_json::Map::new();
+    let mut total_read = 0;
+
+    while total_read < payload_size {
+        // Read the key (must be a valid JSONB text type)
+        let (key_value, key_consumed) = read_jsonb_value(&bytes[total_read..])?;
+        let key_str = key_value
+            .as_str()
+            .ok_or("Invalid object key in JSONB, must be a string")?
+            .to_string();
+        total_read += key_consumed;
+
+        // Read the value associated with the key
+        let (value, value_consumed) = read_jsonb_value(&bytes[total_read..])?;
+        object.insert(key_str, value);
+        total_read += value_consumed;
+    }
+
+    // Ensure the total bytes read match the payload size
+    if total_read != payload_size {
+        return Err("Object payload size mismatch".into());
+    }
+
+    Ok(serde_json::Value::Object(object))
+}
+
+// Helper function to create the correct JsonbHeader based on the payload size
+fn create_jsonb_header(element_type: u8, payload_size: usize) -> Result<Vec<u8>, String> {
+    // Check if payload size exceeds the maximum allowed size
+    if payload_size > 2_147_483_647 {
+        return Err("Payload size exceeds the maximum allowed size of 2GB".into());
+    }
+
+    let mut header = Vec::new();
```

Comment:
We likely should use `Vec::with_capacity()` here as we know the payload size beforehand?